File: app/src/main/java/com/arusoft/mobile/searchlocations/data/network/ApiResponse.java
package com.arusoft.mobile.searchlocations.data.network;

import androidx.annotation.Nullable;

import static com.arusoft.mobile.searchlocations.data.network.HttpErrorCodes.NOT_FOUND_ERROR;


/**
 * Common class used by API responses.
 *
 * @param <R, E>
 */
public class ApiResponse<R, E> {
    @Nullable
    public R body;
    @Nullable
    public E errorBody;
    public Throwable exception;
    public int code;

    public ApiResponse(Throwable error) {
        code = NOT_FOUND_ERROR;
        body = null;
        errorBody = null;
        exception = error;
    }

    public ApiResponse(Response<R, E> response) {
        code = response.code();
        body = response.body();
        exception = null;

        if (isSuccessful()) {
            errorBody = null;
        } else {
            try {
                errorBody = response.error();
            } catch (Exception ex) {
                errorBody = null;
                exception = ex;
            }
        }
    }

    public boolean isSuccessful() {
        return code >= HttpErrorCodes.SUCCESS && code < HttpErrorCodes.REDIRECTION;
    }
}

File: app/src/main/java/com/arusoft/mobile/searchlocations/data/network/ApiResponse.kt
package com.arusoft.mobile.searchlocations.data.network

import com.arusoft.mobile.searchlocations.data.network.HttpErrorCodes.NOT_FOUND_ERROR


/**
 * Common class used by API responses.
 *
 * @param <R></R>, E>
 */
class ApiResponse<R, E> {
    val body: R?
    var errorBody: E? = null
    var exception: Throwable? = null
    var code: Int = 0

    val isSuccessful: Boolean
        get() = code >= HttpErrorCodes.SUCCESS && code < HttpErrorCodes.REDIRECTION

    constructor(error: Throwable) {
        code = NOT_FOUND_ERROR
        body = null
        errorBody = null
        exception = error
    }

    constructor(response: Response<R, E>) {
        code = response.code()
        body = response.body()
        exception = null

        if (isSuccessful) {
            errorBody = null
        } else {
            try {
                errorBody = response.error()
            } catch (ex: Exception) {
                errorBody = null
                exception = ex
            }

        }
    }
}

